Feature/add component tests

HeJones opened the pull request:
### What

To enable features tests to be added to this service the following things have been done: 
- Add component tests to Make File
- Add features package to hold feature tests
- Added example tests to test the hello world endpoint - one is with a custom step, and one uses a step in the library
- Add ci files

### How to review

If you run `make test-component` the feature tests pass

### Who can review
Anyone

HeJones committed:
Add component test library

jon-acker committed:
Add sample component test

jon-acker committed:
Add example of using predefined Then step

mwnicks committed:
add ci component test files

nshumoogum commented on Makefile:
```diff
@@ -38,3 +38,7 @@ test:
+
+.PHONY: test-component
+test-component:
+	go test -cover -coverpkg=github.com/ONSdigital/dp-identity-api/... -component
```
Probably make sense to add -race flag after -cover?

nshumoogum commented on ci/component.yml:
```diff
@@ -0,0 +1,15 @@
+  type: docker-image
+  source:
+    repository: golang
+    tag: 1.15.8
```
Bump to 1.16.3?

nshumoogum replied:
Probably want to bump tagged version in unit.yml and build.yml

nshumoogum commented on go.mod:
```diff
@@ -4,10 +4,12 @@ go 1.15
 
```
go.mod file should be updated to latest go version 1.16

HeJones committed:
Update dependancies

nshumoogum commented on main_test.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func (f *ComponentTest) InitializeTestSuite(ctx *godog.TestSuiteContext) {
+
```
Consider removing empty whitespace

nshumoogum commented on main_test.go:
```diff
@@ -0,0 +1,66 @@
+func (f *ComponentTest) InitializeScenario(ctx *godog.ScenarioContext) {
+	component, err := steps.NewComponent()
+	if err != nil {
+		panic(err)
```
Would it not make more sense to use os.Exit(1)? And include a log line prior to exiting to log out the error?

jon-acker replied:
yes, panic should def be removed, it would stop the test suite :)

nshumoogum commented:
github.com/gogo/protobuf - needs updating to use latest version 1.3.2 instead of 1.3.1 - hopefully this will resolve audit issues

Using `go get -u ./...; go mod tidy`

HeJones committed:
Update go version and dependancies

nshumoogum committed:
Remove second retrieval of environment configurations

Eventhough unlikely this is bad practice as the state of the environment variables
could have changed between the two retrievals, resulting in the app behaving differently
depending on what is being processed.

Instead pass environment variables down the app tree.

nshumoogum committed:
Move initialisation of service in component test suite

Rename files and methods

nshumoogum committed:
Tidy go dependencies

nshumoogum committed:
Increase the time for the service to successfully shutdown in unit test

Currently results in intermittent failure as the time difference set in test is to small
(1ms), upping to 9ms.

HeJones committed:
Update dependancy

HeJones committed:
Update dependancies

nshumoogum committed:
Merge branch 'feature/fix-race-condition' into feature/add-component-tests

HeJones committed:
Tidy dependencies

nshumoogum committed:
Fix up dependencies

mike-kelly-methods commented on features/helloworld.feature:
```diff
@@ -0,0 +1,14 @@
+Feature: Helloworld
+
+
+
```
Excessive number of newlines.

HeJones committed:
Remove additional empty lines

mike-kelly-methods reviewed (approved):
👍
